In [2]:
%%capture
%load_ext sql
%sql sqlite:///chinook.db

'Connected: None@chinook.db'

In [3]:
%%sql

SELECT 
    name,
    type
FROM sqlite_master
WHERE type IN ("table", "view")

Done.


[('album', 'table'),
 ('artist', 'table'),
 ('customer', 'table'),
 ('employee', 'table'),
 ('genre', 'table'),
 ('invoice', 'table'),
 ('invoice_line', 'table'),
 ('media_type', 'table'),
 ('playlist', 'table'),
 ('playlist_track', 'table'),
 ('track', 'table')]

In [4]:
%%sql

SELECT *
FROM invoice

Done.


[(1, 18, '2017-01-03 00:00:00', '627 Broadway', 'New York', 'NY', 'USA', '10012-2612', 15.84),
 (2, 30, '2017-01-03 00:00:00', '230 Elgin Street', 'Ottawa', 'ON', 'Canada', 'K2P 1L7', 9.9),
 (3, 40, '2017-01-05 00:00:00', '8, Rue Hanovre', 'Paris', 'None', 'France', '75002', 1.98),
 (4, 18, '2017-01-06 00:00:00', '627 Broadway', 'New York', 'NY', 'USA', '10012-2612', 7.92),
 (5, 27, '2017-01-07 00:00:00', '1033 N Park Ave', 'Tucson', 'AZ', 'USA', '85719', 16.83),
 (6, 31, '2017-01-10 00:00:00', '194A Chain Lake Drive', 'Halifax', 'NS', 'Canada', 'B3S 1C5', 1.98),
 (7, 49, '2017-01-12 00:00:00', 'Ordynacka 10', 'Warsaw', 'None', 'Poland', '00-358', 10.89),
 (8, 59, '2017-01-13 00:00:00', '3,Raj Bhavan Road', 'Bangalore', 'None', 'India', '560001', 9.9),
 (9, 18, '2017-01-18 00:00:00', '627 Broadway', 'New York', 'NY', 'USA', '10012-2612', 8.91),
 (10, 31, '2017-01-18 00:00:00', '194A Chain Lake Drive', 'Halifax', 'NS', 'Canada', 'B3S 1C5', 1.98),
 (11, 38, '2017-01-20 00:00:00', 'Barbarossastraße 19', 'Berlin', 'None', 'Germany', '10779', 10.89),
 (12, 42, '2017-01-21 00:00:00', '9, Place Louis Barthou', 'Bordeaux', 'None', 'France', '33000', 3.96),
 (13, 35, '2017-01-22 00:00:00', 'Rua dos Campeões Europeus de Viena, 4350', 'Porto', 'None', 'Portugal', 'None', 0.99),
 (14, 25, '2017-01-23 00:00:00', '319 N. Frances Street', 'Madison', 'WI', 'USA', '53703', 0.99),
 (15, 44, '2017-01-25 00:00:00', 'Porthaninkatu 9', 'Helsinki', 'None', 'Finland', '00530', 3.96),
 (16, 1, '2017-01-26 00:00:00', 'Av. Brigadeiro Faria Lima, 2170', 'São José dos Campos', 'SP', 'Brazil', '12227-000', 8.91),
 (17, 20, '2017-01-28 00:00:00', '541 Del Medio Avenue', 'Mountain View', 'CA', 'USA', '94040-111', 10.89),
 (18, 24, '2017-02-02 00:00:00', '162 E Superior Street', 'Chicago', 'IL', 'USA', '60611', 4.95),
 (19, 10, '2017-02-06 00:00:00', 'Rua Dr. Falcão Filho, 155', 'São Paulo', 'SP', 'Brazil', '01007-010', 6.93),
 (20, 43, '2017-02-06 00:00:00', '68, Rue Jouvence', 'Dijon', 'None', 'France', '21000', 9.9),
 (21, 9, '2017-02-07 00:00:00', 'Sønder Boulevard 51', 'Copenhagen', 'None', 'Denmark', '1720', 8.91),
 (22, 35, '2017-02-08 00:00:00', 'Rua dos Campeões Europeus de Viena, 4350', 'Porto', 'None', 'Portugal', 'None', 1.98),
 (23, 53, '2017-02-09 00:00:00', '113 Lupus St', 'London', 'None', 'United Kingdom', 'SW1V 3EN', 9.9),
 (24, 4, '2017-02-10 00:00:00', 'Ullevålsveien 14', 'Oslo', 'None', 'Norway', '0171', 11.879999999999999),
 (25, 39, '2017-02-11 00:00:00', '4, Rue Milton', 'Paris', 'None', 'France', '75009', 7.92),
 (26, 55, '2017-02-15 00:00:00', '421 Bourke Street', 'Sidney', 'NSW', 'Australia', '2010', 1.98),
 (27, 35, '2017-02-15 00:00:00', 'Rua dos Campeões Europeus de Viena, 4350', 'Porto', 'None', 'Portugal', 'None', 7.92),
 (28, 12, '2017-02-17 00:00:00', 'Praça Pio X, 119', 'Rio de Janeiro', 'RJ', 'Brazil', '20040-020', 2.9699999999999998),
 (29, 45, '2017-02-17 00:00:00', 'Erzsébet krt. 58.', 'Budapest', 'None', 'Hungary', 'H-1073', 5.9399999999999995),
 (30, 46, '2017-02-18 00:00:00', '3 Chatham Street', 'Dublin', 'Dublin', 'Ireland', 'None', 10.89),
 (31, 3, '2017-02-21 00:00:00', '1498 rue Bélanger', 'Montréal', 'QC', 'Canada', 'H2G 1A7', 19.8),
 (32, 30, '2017-02-21 00:00:00', '230 Elgin Street', 'Ottawa', 'ON', 'Canada', 'K2P 1L7', 12.87),
 (33, 58, '2017-02-21 00:00:00', '12,Community Centre', 'Delhi', 'None', 'India', '110017', 3.96),
 (34, 33, '2017-02-23 00:00:00', '5112 48 Street', 'Yellowknife', 'NT', 'Canada', 'X1A 1N6', 0.99),
 (35, 34, '2017-02-25 00:00:00', 'Rua da Assunção 53', 'Lisbon', 'None', 'Portugal', 'None', 4.95),
 (36, 51, '2017-02-27 00:00:00', 'Celsiusg. 9', 'Stockholm', 'None', 'Sweden', '11230', 6.93),
 (37, 59, '2017-03-05 00:00:00', '3,Raj Bhavan Road', 'Bangalore', 'None', 'India', '560001', 9.9),
 (38, 49, '2017-03-05 00:00:00', 'Ordynacka 10', 'Warsaw', 'None', 'Poland', '00-358', 6.93),
 (39, 39, '2017-03-12 00:00:00', '4, Rue Milton', 'Paris', 'None', 'France', '75009', 7.92),
 (40, 37, '2017

In [5]:
%%sql

SELECT *
FROM customer

Done.


[(1, 'Luís', 'Gonçalves', 'Embraer - Empresa Brasileira de Aeronáutica S.A.', 'Av. Brigadeiro Faria Lima, 2170', 'São José dos Campos', 'SP', 'Brazil', '12227-000', '+55 (12) 3923-5555', '+55 (12) 3923-5566', 'luisg@embraer.com.br', 3),
 (2, 'Leonie', 'Köhler', None, 'Theodor-Heuss-Straße 34', 'Stuttgart', None, 'Germany', '70174', '+49 0711 2842222', None, 'leonekohler@surfeu.de', 5),
 (3, 'François', 'Tremblay', None, '1498 rue Bélanger', 'Montréal', 'QC', 'Canada', 'H2G 1A7', '+1 (514) 721-4711', None, 'ftremblay@gmail.com', 3),
 (4, 'Bjørn', 'Hansen', None, 'Ullevålsveien 14', 'Oslo', None, 'Norway', '0171', '+47 22 44 22 22', None, 'bjorn.hansen@yahoo.no', 4),
 (5, 'František', 'Wichterlová', 'JetBrains s.r.o.', 'Klanova 9/506', 'Prague', None, 'Czech Republic', '14700', '+420 2 4172 5555', '+420 2 4172 5555', 'frantisekw@jetbrains.com', 4),
 (6, 'Helena', 'Holý', None, 'Rilská 3174/6', 'Prague', None, 'Czech Republic', '14300', '+420 2 4177 0449', None, 'hholy@gmail.com', 5),
 (7, 'Astrid', 'Gruber', None, 'Rotenturmstraße 4, 1010 Innere Stadt', 'Vienne', None, 'Austria', '1010', '+43 01 5134505', None, 'astrid.gruber@apple.at', 5),
 (8, 'Daan', 'Peeters', None, 'Grétrystraat 63', 'Brussels', None, 'Belgium', '1000', '+32 02 219 03 03', None, 'daan_peeters@apple.be', 4),
 (9, 'Kara', 'Nielsen', None, 'Sønder Boulevard 51', 'Copenhagen', None, 'Denmark', '1720', '+453 3331 9991', None, 'kara.nielsen@jubii.dk', 4),
 (10, 'Eduardo', 'Martins', 'Woodstock Discos', 'Rua Dr. Falcão Filho, 155', 'São Paulo', 'SP', 'Brazil', '01007-010', '+55 (11) 3033-5446', '+55 (11) 3033-4564', 'eduardo@woodstock.com.br', 4),
 (11, 'Alexandre', 'Rocha', 'Banco do Brasil S.A.', 'Av. Paulista, 2022', 'São Paulo', 'SP', 'Brazil', '01310-200', '+55 (11) 3055-3278', '+55 (11) 3055-8131', 'alero@uol.com.br', 5),
 (12, 'Roberto', 'Almeida', 'Riotur', 'Praça Pio X, 119', 'Rio de Janeiro', 'RJ', 'Brazil', '20040-020', '+55 (21) 2271-7000', '+55 (21) 2271-7070', 'roberto.almeida@riotur.gov.br', 3),
 (13, 'Fernanda', 'Ramos', None, 'Qe 7 Bloco G', 'Brasília', 'DF', 'Brazil', '71020-677', '+55 (61) 3363-5547', '+55 (61) 3363-7855', 'fernadaramos4@uol.com.br', 4),
 (14, 'Mark', 'Philips', 'Telus', '8210 111 ST NW', 'Edmonton', 'AB', 'Canada', 'T6G 2C7', '+1 (780) 434-4554', '+1 (780) 434-5565', 'mphilips12@shaw.ca', 5),
 (15, 'Jennifer', 'Peterson', 'Rogers Canada', '700 W Pender Street', 'Vancouver', 'BC', 'Canada', 'V6C 1G8', '+1 (604) 688-2255', '+1 (604) 688-8756', 'jenniferp@rogers.ca', 3),
 (16, 'Frank', 'Harris', 'Google Inc.', '1600 Amphitheatre Parkway', 'Mountain View', 'CA', 'USA', '94043-1351', '+1 (650) 253-0000', '+1 (650) 253-0000', 'fharris@google.com', 4),
 (17, 'Jack', 'Smith', 'Microsoft Corporation', '1 Microsoft Way', 'Redmond', 'WA', 'USA', '98052-8300', '+1 (425) 882-8080', '+1 (425) 882-8081', 'jacksmith@microsoft.com', 5),
 (18, 'Michelle', 'Brooks', None, '627 Broadway', 'New York', 'NY', 'USA', '10012-2612', '+1 (212) 221-3546', '+1 (212) 221-4679', 'michelleb@aol.com', 3),
 (19, 'Tim', 'Goyer', 'Apple Inc.', '1 Infinite Loop', 'Cupertino', 'CA', 'USA', '95014', '+1 (408) 996-1010', '+1 (408) 996-1011', 'tgoyer@apple.com', 3),
 (20, 'Dan', 'Miller', None, '541 Del Medio Avenue', 'Mountain View', 'CA', 'USA', '94040-111', '+1 (650) 644-3358', None, 'dmiller@comcast.com', 4),
 (21, 'Kathy', 'Chase', None, '801 W 4th Street', 'Reno', 'NV', 'USA', '89503', '+1 (775) 223-7665', None, 'kachase@hotmail.com', 5),
 (22, 'Heather', 'Leacock', None, '120 S Orange Ave', 'Orlando', 'FL', 'USA', '32801', '+1 (407) 999-7788', None, 'hleacock@gmail.com', 4),
 (23, 'John', 'Gordon', None, '69 Salem Street', 'Boston', 'MA', 'USA', '2113', '+1 (617) 522-1333', None, 'johngordon22@yahoo.com', 4),
 (24, 'Frank', 'Ralston', None, '162 E Superior Street', 'Chicago', 'IL', 'USA', '60611', '+1 (312) 332-3232', None, 'fralston@gmail.com', 3),
 (25, 'Victor', 'Stevens', None, '319 N. Frances Street', 'Madison', 'WI', 'USA', '53703', '+1 (608) 257-0597', N

In [6]:
%%sql

SELECT
    g.name AS genre,
    COUNT(t.track_id) AS number_of_tracks
FROM genre AS g
LEFT JOIN track AS t ON t.genre_id = g.genre_id
GROUP BY 1
ORDER BY 2 DESC

Done.


[('Rock', 1297),
 ('Latin', 579),
 ('Metal', 374),
 ('Alternative & Punk', 332),
 ('Jazz', 130),
 ('TV Shows', 93),
 ('Blues', 81),
 ('Classical', 74),
 ('Drama', 64),
 ('R&B/Soul', 61),
 ('Reggae', 58),
 ('Pop', 48),
 ('Soundtrack', 43),
 ('Alternative', 40),
 ('Hip Hop/Rap', 35),
 ('Electronica/Dance', 30),
 ('Heavy Metal', 28),
 ('World', 28),
 ('Sci Fi & Fantasy', 26),
 ('Easy Listening', 24),
 ('Comedy', 17),
 ('Bossa Nova', 15),
 ('Science Fiction', 13),
 ('Rock And Roll', 12),
 ('Opera', 1)]

In [7]:
%%sql

WITH 
    tracks_usa AS
               (
                SELECT
                    t.track_id,
                    t.genre_id,
                    g.name AS genre,
                    i.billing_country,
                    i.invoice_id,
                    t.album_id,
                    ar.name AS artist_name
                FROM track AS t
                INNER JOIN genre AS g ON g.genre_id = t.genre_id
                INNER JOIN album AS a ON a.album_id = t.album_id
                INNER JOIN artist AS ar ON ar.artist_id = a.artist_id
                INNER JOIN invoice_line AS il ON il.track_id = t.track_id
                INNER JOIN invoice AS i ON i.invoice_id = il.invoice_id
                WHERE i.billing_country = 'USA'
               )
SELECT 
    tusa.genre,
    COUNT(tusa.invoice_id) AS number_of_purchases,
    COUNT(tusa.invoice_id) / (
                             SELECT
                                 CAST(COUNT(invoice_id) AS FLOAT)
                             FROM tracks_usa 
                           ) * 100 AS number_of_purchases_per
FROM tracks_usa AS tusa
GROUP BY 1
ORDER BY 3 DESC

Done.


[('Rock', 561, 53.37773549000951),
 ('Alternative & Punk', 130, 12.369172216936251),
 ('Metal', 124, 11.798287345385347),
 ('R&B/Soul', 53, 5.042816365366318),
 ('Blues', 36, 3.425309229305423),
 ('Alternative', 35, 3.3301617507136063),
 ('Latin', 22, 2.093244529019981),
 ('Pop', 22, 2.093244529019981),
 ('Hip Hop/Rap', 20, 1.9029495718363463),
 ('Jazz', 14, 1.3320647002854424),
 ('Easy Listening', 13, 1.236917221693625),
 ('Reggae', 6, 0.570884871550904),
 ('Electronica/Dance', 5, 0.47573739295908657),
 ('Classical', 4, 0.3805899143672693),
 ('Heavy Metal', 3, 0.285442435775452),
 ('Soundtrack', 2, 0.19029495718363465),
 ('TV Shows', 1, 0.09514747859181732)]

The most popular genre purchased in the **USA** is the **Rock** genre. Out of track purchases in the US, it makes up **53%** of purchases. The other two popular genres are **Alternative & Punk** and **Metal**.

In [8]:
%%sql

WITH 
    tracks_usa AS
               (
                
                SELECT
                    t.track_id,
                    t.genre_id,
                    g.name AS genre,
                    i.billing_country,
                    i.invoice_id,
                    t.album_id,
                    a.title AS album_name,
                    ar.name AS artist_name
                FROM track AS t
                INNER JOIN genre AS g ON g.genre_id = t.genre_id
                INNER JOIN album AS a ON a.album_id = t.album_id
                INNER JOIN artist AS ar ON ar.artist_id = a.artist_id
                INNER JOIN invoice_line AS il ON il.track_id = t.track_id
                INNER JOIN invoice AS i ON i.invoice_id = il.invoice_id
                WHERE i.billing_country = 'USA'
               )
SELECT 
    tusa.artist_name AS artist_name,
    tusa.album_name AS album_name,
    tusa.genre AS genre,
    COUNT(tusa.invoice_id) AS number_of_purchases,
    COUNT(tusa.invoice_id) / (
                             SELECT
                                 CAST(COUNT(invoice_id) AS FLOAT)
                             FROM tracks_usa 
                           ) * 100 AS number_of_purchases_per
FROM tracks_usa AS tusa
WHERE genre = 'Metal'
GROUP BY 2
ORDER BY 4 DESC

Done.


[('System Of A Down', 'Mezmerize', 'Metal', 22, 2.093244529019981),
 ('Godsmack', 'Faceless', 'Metal', 20, 1.9029495718363463),
 ('Black Sabbath', 'Black Sabbath', 'Metal', 18, 1.7126546146527115),
 ("Guns N' Roses", 'Use Your Illusion II', 'Metal', 9, 0.8563273073263558),
 ('Judas Priest', 'Living After Midnight', 'Metal', 8, 0.7611798287345386),
 ('Apocalyptica', 'Plays Metallica By Four Cellos', 'Metal', 8, 0.7611798287345386),
 ('Black Sabbath', 'Black Sabbath Vol. 4 (Remaster)', 'Metal', 6, 0.570884871550904),
 ('Motörhead', 'Ace Of Spades', 'Metal', 5, 0.47573739295908657),
 ('Mötley Crüe', 'Motley Crue Greatest Hits', 'Metal', 5, 0.47573739295908657),
 ('Metallica', 'Garage Inc. (Disc 2)', 'Metal', 3, 0.285442435775452),
 ('Lenny Kravitz', 'Greatest Hits', 'Metal', 3, 0.285442435775452),
 ('Metallica', 'Load', 'Metal', 3, 0.285442435775452),
 ('Metallica', 'ReLoad', 'Metal', 3, 0.285442435775452),
 ('Ozzy Osbourne', 'Tribute', 'Metal', 3, 0.285442435775452),
 ('Iron Maiden', 'A Real Live One', 'Metal', 2, 0.19029495718363465),
 ('Metallica', '...And Justice For All', 'Metal', 1, 0.09514747859181732),
 ('Black Label Society', 'Alcohol Fueled Brewtality Live! [Disc 1]', 'Metal', 1, 0.09514747859181732),
 ('Bruce Dickinson', 'Chemical Wedding', 'Metal', 1, 0.09514747859181732),
 ('Metallica', 'Master Of Puppets', 'Metal', 1, 0.09514747859181732),
 ('Metallica', 'Ride The Lightning', 'Metal', 1, 0.09514747859181732),
 ('Iron Maiden', 'Somewhere in Time', 'Metal', 1, 0.09514747859181732)]

The most popular artist purchased out of the rock genre is **Van Halen** with the album **The Best Of Van Halen, Vol. I**. The most purchased album in the genre is **Are You Experienced?** by **Jimi Hendrix**.

The artist with the most purchases in the **Alternative & Punk** is **Green Day** with the album **American Idiot**. The most purchased album is **Green** by **R.E.M.**

The artist with the most purchases in the **Metal** genre is **Black Sabbath** with the album **Black Sabbath**. The most purchased album is **Mezmerize** by **System Of A Down**.


In [9]:
%%sql

WITH employee_info AS 
                    (
                     SELECT 
                         c.customer_id,
                         c.support_rep_id,
                         c.first_name || ' ' || c.last_name AS customer_name,
                         e.employee_id,
                         e.hire_date,
                         e.first_name || ' ' || e.last_name AS employee_name
                     FROM customer AS c
                     INNER JOIN employee AS e ON e.employee_id = c.support_rep_id
                    )
SELECT
    ei.employee_name,
    ei.hire_date,
    COUNT(i.invoice_id) AS number_sales,
    CAST(SUM(i.total) AS float) AS total_sales
FROM employee_info AS ei
LEFT JOIN invoice AS i ON i.customer_id = ei.customer_id
GROUP BY 1

Done.


[('Jane Peacock', '2017-04-01 00:00:00', 212, 1731.510000000004),
 ('Margaret Park', '2017-05-03 00:00:00', 214, 1584.0000000000032),
 ('Steve Johnson', '2017-10-17 00:00:00', 188, 1393.9200000000028)]

In [10]:
%%sql

WITH country_or_other AS
    (
     SELECT
       CASE
           WHEN (
                 SELECT count(*)
                 FROM customer
                 where country = c.country
                ) = 1 THEN "Other"
           ELSE c.country
       END AS country,
       c.customer_id,
       il.*
     FROM invoice_line il
     INNER JOIN invoice i ON i.invoice_id = il.invoice_id
     INNER JOIN customer c ON c.customer_id = i.customer_id
    )

SELECT
    country,
    customers,
    total_sales,
    average_order,
    customer_lifetime_value
FROM
    (
    SELECT
        country,
        count(distinct customer_id) customers,
        SUM(unit_price) total_sales,
        SUM(unit_price) / count(distinct customer_id) customer_lifetime_value,
        SUM(unit_price) / count(distinct invoice_id) average_order,
        CASE
            WHEN country = "Other" THEN 1
            ELSE 0
        END AS sort
    FROM country_or_other
    GROUP BY country
    ORDER BY sort ASC, total_sales DESC
    );

Done.


[('USA', 13, 1040.490000000008, 7.942671755725252, 80.03769230769292),
 ('Canada', 8, 535.5900000000034, 7.047236842105309, 66.94875000000043),
 ('Brazil', 5, 427.68000000000245, 7.011147540983647, 85.53600000000048),
 ('France', 5, 389.0700000000021, 7.781400000000042, 77.81400000000042),
 ('Germany', 4, 334.6200000000016, 8.161463414634186, 83.6550000000004),
 ('Czech Republic', 2, 273.24000000000103, 9.108000000000034, 136.62000000000052),
 ('United Kingdom', 3, 245.52000000000078, 8.768571428571457, 81.84000000000026),
 ('Portugal', 2, 185.13000000000022, 6.3837931034482835, 92.56500000000011),
 ('India', 2, 183.1500000000002, 8.72142857142858, 91.5750000000001),
 ('Other', 15, 1094.9400000000085, 7.448571428571486, 72.99600000000056)]

In [11]:
%%sql

WITH invoice_first_track AS
    (
     SELECT
         il.invoice_id invoice_id,
         MIN(il.track_id) first_track_id
     FROM invoice_line il
     GROUP BY 1
    )

SELECT
    album_purchase,
    COUNT(invoice_id) number_of_invoices,
    CAST(count(invoice_id) AS FLOAT) / (
                                         SELECT COUNT(*) FROM invoice
                                      ) percent
FROM
    (
    SELECT
        ifs.*,
        CASE
            WHEN
                 (
                  SELECT t.track_id FROM track t
                  WHERE t.album_id = (
                                      SELECT t2.album_id FROM track t2
                                      WHERE t2.track_id = ifs.first_track_id
                                     ) 

                  EXCEPT 

                  SELECT il2.track_id FROM invoice_line il2
                  WHERE il2.invoice_id = ifs.invoice_id
                 ) IS NULL
             AND
                 (
                  SELECT il2.track_id FROM invoice_line il2
                  WHERE il2.invoice_id = ifs.invoice_id

                  EXCEPT 

                  SELECT t.track_id FROM track t
                  WHERE t.album_id = (
                                      SELECT t2.album_id FROM track t2
                                      WHERE t2.track_id = ifs.first_track_id
                                     ) 
                 ) IS NULL
             THEN "yes"
             ELSE "no"
         END AS "album_purchase"
     FROM invoice_first_track ifs
    )
GROUP BY album_purchase;

Done.


[('no', 500, 0.8143322475570033), ('yes', 114, 0.18566775244299674)]

Album purchases account for 18.6% of purchases. Based on this data, I would recommend against purchasing only select tracks from albums from record companies, since there is potential to lose one fifth of revenue.